### PR TITLE
add the server error client error custome error

### PR DESCRIPTION
**Pull request by momenmarwan**
add server error and client error custom error middleware

### CODE REVIEW

**Review by Farah202324 — APPROVED** [review]

**Review by momenqudaih — DISMISSED** [review]

File: server/middleware/clientError.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { Request, Response } from 'express';
```
Comment (moustf) [inline server/middleware/clientError.ts:1]: Add a new line between the imports and the function.

File: server/middleware/clientError.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+const clientError = (req: Request, res: Response) => {
+  res.status(404).json({
+    error: true,
+    massage: 'page not found',
```
Comment (moustf) [inline server/middleware/clientError.ts:5]: Edit the typo; message.

File: server/middleware/serverError.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+const serverError = (
+  err: unknown,
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  if (err instanceof ValidationError) {
+    return res.status(400).json({
+      status: 401,
```
Comment (moustf) [inline server/middleware/serverError.ts:15]: Wrong data response is 400, and you don't have to include it in the response.

File: server/middleware/serverError.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { CustomError } from '../utils';
+import { NextFunction, Request, Response } from 'express';
+
+const serverError = (
+  err: unknown,
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  if (err instanceof ValidationError) {
```
Comment (moustf) [inline server/middleware/serverError.ts:13]: When you throw a new CustomError each time you face an error in your controller, you don't need these checks.
You have to decide on a paradigm to use with your team.

File: server/utils/helper/customError.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export default class CustomError extends Error {
+  constructor(public status: number, msg: string) {
+    super(msg);
```
Comment (moustf) [inline server/utils/helper/customError.ts:3]: add
```ts
this.status = status;
```

**Review by moustf — CHANGES_REQUESTED** [review]
Please refer to my last message on the Discord channel for further details.

**Review by Farah202324 — DISMISSED** [review]

**Review by Farah202324 — APPROVED** [review]

**Review by momenqudaih — APPROVED** [review]

**Review by moustf — APPROVED** [review]

**Pull request merged**